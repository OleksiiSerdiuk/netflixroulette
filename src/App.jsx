import Header from './components/Header/Header';
import Footer from './components/Footer/Footer';
import Movies from './components/movies/Movies';
import MovieDetail from './components/Header/components/MovieDetail/MovieDetail';
import { MovieContext } from './contexts';
import { useContext } from 'react';
import { Route, Routes, Navigate } from 'react-router-dom';
import NotFoundPage from './components/NotFoundPage/NotFoundPage';

const App = () => {
  const { movieDescription } = useContext(MovieContext);

  const Main = () => {
    return (
      <>
        {!movieDescription ? <Header /> : <MovieDetail />}
        <Movies />
        <Footer />
      </>
    );
  };

  return (
    <Routes>
      <Route exact path='/' element={<Navigate to='/search' replace />} />
      <Route path='/search' element={<Main />}>
        <Route path=':movie' element={<Main />} />
      </Route>
      <Route
        path='/movie/:id'
        element={
          movieDescription ? <Main /> : <Navigate to='/search' replace />
        }
      />
      <Route path='*' element={<NotFoundPage />} />
    </Routes>
  );
};

export default App;
