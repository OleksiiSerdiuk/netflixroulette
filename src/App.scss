// Override default variables before the import
@import url('https://fonts.googleapis.com/css2?family=Cinzel+Decorative:wght@400;700;900&family=Montserrat:wght@100;200;300;400;500;600;700;800;900&family=Roboto:wght@100;300;400;500;700;900&display=swap');
@import './styles/main.scss';
