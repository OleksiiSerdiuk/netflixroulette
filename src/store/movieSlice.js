import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import API from '../api';

export const initialState = {
  movies: null,
  isLoading: false,
  serverError: null,
  filters: ['all', 'documentary', 'comedy', 'horror', 'crime'],
  sortBy: 'release_date',
  filter: 'all',
  searchField: '',
};

export const fetchMovies = createAsyncThunk(
  '/movies/fetchMovies',
  async (property, { rejectWithValue, getState }) => {
    const {
      movies: { sortBy, filter, searchField },
    } = getState();

    const currentFilter = filter === 'all' ? '' : filter;

    try {
      const response = await API.get(
        `/movies?sortBy=${sortBy}&sortOrder=asc&filter=${currentFilter}&limit=9&searchBy=title&search=${searchField}`
      );

      return await response.data;
    } catch (err) {
      return rejectWithValue(err.response);
    }
  }
);

export const removeMovie = createAsyncThunk(
  '/movies/removeMovies',
  async (movieId, { rejectWithValue }) => {
    try {
      await API.delete(`/movies/${movieId}`);
    } catch (err) {
      return rejectWithValue(err.response);
    }
  }
);

export const createMovie = createAsyncThunk(
  '/movies/createMovie',
  async (data, { rejectWithValue }) => {
    try {
      await API.post(`/movies`, data);
    } catch (err) {
      return rejectWithValue(err.response);
    }
  }
);

export const updateMovie = createAsyncThunk(
  '/movies/updateMovie',
  async (data, { rejectWithValue }) => {
    try {
      await API.put(`/movies`, data);
    } catch (err) {
      return rejectWithValue(err.response);
    }
  }
);

export const movieSlice = createSlice({
  name: 'movies',
  initialState,
  reducers: {
    setFilter: (state, action) => {
      state.filter = action.payload;
    },
    setSort: (state, action) => {
      state.sortBy = action.payload;
    },
    setSearchField: (state, action) => {
      state.searchField = action.payload;
    },
  },
  extraReducers: {
    [fetchMovies.pending]: (state) => {
      state.isLoading = true;
    },
    [fetchMovies.fulfilled]: (state, action) => {
      state.movies = action.payload.data;
      state.isLoading = false;
    },
    [fetchMovies.rejected]: (state, action) => {
      state.serverError = action.payload.data;
      state.isLoading = false;
    },
  },
});

export const { setFilter, setSort, setSearchField } = movieSlice.actions;

export default movieSlice.reducer;
