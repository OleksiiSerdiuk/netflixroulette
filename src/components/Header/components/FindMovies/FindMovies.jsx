import Button from '../../../common/Button';
import { Formik } from 'formik';
import * as Yup from 'yup';
import FormikControl from '../../../Formik/FormikControl';
import { fetchMovies, setSearchField } from '../../../../store/movieSlice';
import { useDispatch, useSelector } from 'react-redux';
import { createSearchParams, useNavigate, useParams } from 'react-router-dom';
import { useEffect, useState } from 'react';

const validationSchema = Yup.object().shape({
  search: Yup.string().min(2),
});

const FindMovies = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { filter, sortBy } = useSelector((state) => state.movies);
  let { movie } = useParams();

  const [query, setQuery] = useState(movie);

  useEffect(() => {
    if (!query) {
      navigate('/search');
      dispatch(fetchMovies());
    } else {
      setTimeout(() => {
        dispatch(setSearchField(query));
        dispatch(fetchMovies());
      });
    }
  }, [query]);

  const handleSearch = (values) => {
    setQuery(values.search);
    dispatch(setSearchField(values.search));
    navigate({
      pathname: `/search/${values.search}`,
      search: `?${createSearchParams({
        genre: filter,
        sort: sortBy,
      })}`,
    });
  };

  return (
    <Formik
      initialValues={{ search: query || '' }}
      onSubmit={(values, { setSubmitting }) => {
        handleSearch(values);
        setSubmitting(false);
      }}
      validationSchema={validationSchema}
    >
      {(props) => {
        const { values, handleChange, handleBlur, handleSubmit } = props;
        return (
          <form action='#' className='find-movie-form' onSubmit={handleSubmit}>
            <FormikControl
              control='input'
              name='search'
              type='text'
              placeholder='What do you want to watch ?'
              customClass='find-movie-form__input'
              value={values.search}
              onChange={handleChange}
              onBlur={handleBlur}
            />
            <Button
              btnClass='find-movie-form__btn'
              btnText='Search'
              type={'submit'}
            />
          </form>
        );
      }}
    </Formik>
  );
};

export default FindMovies;
