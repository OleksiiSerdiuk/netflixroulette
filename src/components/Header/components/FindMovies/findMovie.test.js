import { render, fireEvent, waitFor } from '@testing-library/react';
import FindMovies from './FindMovies';
import store from '../../../../store';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { act } from 'react-dom/test-utils';
import '@testing-library/jest-dom';
import App from '../../../../App';

const handleSubmit = jest.fn();

describe('Search section', () => {
  it('renders find movie snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <FindMovies onSubmit={handleSubmit} />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment(<FindMovies />)).toMatchSnapshot();
  });

  it('should handle submit', async () => {
    const { getByRole } = render(
      <Provider store={store}>
        <BrowserRouter>
          <FindMovies onSubmit={handleSubmit} />
        </BrowserRouter>
      </Provider>
    );

    await act(() => {
      fireEvent.change(getByRole('textbox'), { target: { value: 'test' } });
      fireEvent.click(getByRole('button', { type: /submit/i }));
      handleSubmit();
    });

    await waitFor(() => {
      expect(handleSubmit).toHaveBeenCalledTimes(1);
      expect(getByRole('textbox').value).toBe('test');
    });
  });

  it('should have button', () => {
    const { getByRole } = render(
      <Provider store={store}>
        <BrowserRouter>
          <FindMovies onSubmit={handleSubmit} />
        </BrowserRouter>
      </Provider>
    );
    expect(getByRole('button')).toBeTruthy();
  });

  it('should have value from path', () => {
    window.history.pushState({}, '', '/search/test');

    const { getByRole } = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    console.log('window.history', window.history);
    console.log('document.referrer', document.referrer);
    expect(getByRole('textbox').value).toMatch('test');
  });
  afterAll(() => {});
});
