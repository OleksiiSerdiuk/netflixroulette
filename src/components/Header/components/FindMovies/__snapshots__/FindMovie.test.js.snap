// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search section renders find movie snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="find-movie-form"
  >
    <label
      class="base-input find-movie-form__input"
    >
      <input
        class="base-input__field"
        id="search"
        name="search"
        placeholder="What do you want to watch ?"
        type="text"
        value=""
      />
    </label>
    <button
      class="btn find-movie-form__btn"
      type="submit"
    >
      Search
    </button>
  </form>
</DocumentFragment>
`;
